### ISSUE TITLE
Feature: Forestry tools

### ISSUE THREAD

**Issue by pauldpickell** [ubc-minetest-classroom/Minetest_Classroom#46]
- Long measuring tape (player movement required - may be useful to have a physical tape laid out behind the player as the player walks since this tape is also used to do transect sampling: logs/twigs/debris that intersect the tape are counted or measured in some way)
- Rangefinder (basically a digital measuring device that uses a laser - no player movement required)
- Clinometer (used together with either a rangefinder or long measuring tape to measure the angle
- Compass (used for wayfinding, need to consider implementing magnetic declination setting by the player)
- Diameter (D-)tape (used to measure the diameter [at breast height] of a tree - not sure how well this translates to Minetest since current tree trunks do not vary in diameter)
- Wedge prism ([see this cool video](https://www.youtube.com/watch?v=q_3d6UhSUC8))
- Plot centre pole (used to mark the centre of a plot for either fixed or variable sampling design)
- Flagging tape (used to wrap around trees to identify the boundary of a circular plot or the plot centre)

**Comment by pauldpickell** [comment]
These should probably be organized into a stand alone `forestry-tools` mod that is not dependent on `Minetest_Classroom`.

**Comment by pauldpickell** [comment]
From @lukasdragon in [idea card](https://github.com/ubc-minetest-classroom/Minetest_Classroom/projects/1#card-82505646):

> Student tools (clinometer, compass, meter tape, DBH tape); Being able to teach students how these tools work before field school is important so that instructors can focus on more important topics and so that students have a working understanding of the tools before being asked to use them under time-sensitive contexts.

**Comment by jocelynbachmann** [comment]
As an update, we’ve got the basic functionality working for the compass and measuring tape tools, but still have some work to do on aesthetics. For the compass, it currently just updates the image in the inventory and in-hand each time the direction changes, but we’re not sure if that’s the best way to display it – let us know if you had something else in mind. For the measuring tape, I plan to have it lay out a measuring tape between the two points the player sets (currently just changes the start and end blocks). If there’s any game artist that’s free it would be super helpful to get a texture for the actual tape (like a yellow/black line through the centre of a 16x16 png), if not no worries I can make it, just might look a little wonky.

**Comment by pauldpickell** [comment]
For the compass, it would be useful if we could display an image of what the compass looks like when you hold it at eye level as opposed to just wielding so that we can give the player more of the experience with way finding. Something like this HUD element from the [MT-CTF rifle scoping experience](https://github.com/MT-CTF/capturetheflag/blob/3ec8c740479749d1ac8cc48cdf263800c596a04a/mods/ctf/ctf_combat/ctf_ranged/init.lua#L209-L215):
```
hud:add(player, "ctf_ranged:scope", {
	hud_elem_type = "image",
	position = {x = 0.5, y = 0.5},
	text = "ctf_ranged_rifle_crosshair.png",
	scale = {x = scale_const, y = scale_const},
	alignment = {x = "center", y = "center"},
})
```
There are other elements that I think are needed, like a formspec to capture the magnetic declination, something interactive to set a bearing, some way to put "red in the shed". All this might become more clear after you work with a compass.
Edit: Also, we should try to avoid floating point degrees. Let's stick with integers.

**Comment by pauldpickell** [comment]
Just noticed behaviour on the measuring tape where the yellow node placed on-click persists after leaving the game. Probably need to add an on_leaveplayer callback to make sure that gets cleared.

**Comment by jocelynbachmann** [comment]
Measuring tape fixed in #126

**Comment by pauldpickell** [comment]
I am fine with the tools that have been completed so far. Other bugs that have come up for some tools have already been identified in other issues, so I am closing this for now.
